utils/yoke/src/yokeable.rs

```diff
@@ -0,0 +1,62 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use std::borrow::{Cow, ToOwned};
+use std::mem;
+
+pub unsafe trait Yokeable<'a>: 'static {
+    type Output: 'a + Sized; // MUST be `Self` with the lifetime swapped
+                             // used by `SharedData::get()`
+    fn transform(&'a self) -> &'a Self::Output;
+
+    // Used for zero-copy deserialization. Safety constraint: `Self` must be
+    // destroyed before the data `Self::Output` was deserialized
+    // from is
+    unsafe fn make(from: Self::Output) -> Self;
+
+    fn with_mut<F>(&'a mut self, f: F)
+    where
+        F: 'static + FnOnce(&'a mut Self::Output);
+}
+
+unsafe impl<'a, T: 'static + ToOwned + ?Sized> Yokeable<'a> for Cow<'static, T>
+where
+    <T as ToOwned>::Owned: Sized,
+{
+    type Output = Cow<'a, T>;
+    fn transform(&'a self) -> &'a Cow<'a, T> {
+        self
+    }
+
+    unsafe fn make(from: Cow<'a, T>) -> Self {
+        debug_assert!(mem::size_of::<Cow<'a, T>>() == mem::size_of::<Self>());
```

Comment:
static_assert instead?

Reply:
rust doesn't have that; because Rust generics are not typechecked at substitution (unlike templates)

Reply:
So, in other words, this is impossible to test at build time, even if you were to use https://docs.rs/static_assertions/1.1.0/static_assertions/ ?

Reply:
Yes, and the static assertions crate uses tricks like `mem::transmute` which doesn't work here.

It might be possible to test using `const`s. I'm not sure, but it would be super happy